lib/Dialect/Moore/Transforms/LowerConcatRef.cpp

```diff
@@ -0,0 +1,128 @@
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_LOWERCONCATREF
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+using namespace mlir;
+
+namespace {
+
+// struct ConcatRefLowering : public OpConversionPattern<ConcatRefOp> {
+//   using OpConversionPattern::OpConversionPattern;
+
+//   LogicalResult
+//   matchAndRewrite(ConcatRefOp op, OpAdaptor adaptor,
+//                   ConversionPatternRewriter &rewriter) const final {
+//     auto *user = *op->getUsers().begin();
+
+//     return success();
+//   }
+// };
+
+// template<typename OpTy>
+struct ConcatRefLowering : public OpConversionPattern<ConcatRefOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ConcatRefOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    if (auto *user = *op->getUsers().begin()) {
+      auto src = user->getOperand(1);
+      auto srcWidth = cast<UnpackedType>(src.getType()).getBitSize().value();
+
+      for (auto operand : op.getValues()) {
+        auto type = cast<RefType>(operand.getType()).getNestedType();
+        auto width = type.getBitSize().value();
+
+        rewriter.setInsertionPoint(user);
+        auto i32 = moore::IntType::getInt(getContext(), 32);
+        auto lowBit =
+            rewriter.create<ConstantOp>(op.getLoc(), i32, srcWidth - width);
+        auto extract =
+            rewriter.create<ExtractOp>(src.getLoc(), type, src, lowBit);
+
+        // Update the real bit width of RHS of assignment. Like "c" the above
+        // description mentioned.
+        srcWidth = srcWidth - width;
+
+        // Estimate which assignment needs to be created.
+
+        rewriter.create<BlockingAssignOp>(user->getLoc(), operand, extract);
+      }
+      rewriter.eraseOp(user);
+      rewriter.eraseOp(op);
+    }
+    return success();
+  }
+};
+
+// template<typename OpTy>
+// struct AssignOpLowering : public
+// OpConversionPattern<BlockingAssignOp> {
+//   using OpConversionPattern::OpConversionPattern;
+
+//   LogicalResult
+//   matchAndRewrite(BlockingAssignOp op, OpAdaptor adaptor,
+//                   ConversionPatternRewriter &rewriter) const final {
+//     rewriter.eraseOp(op);
+//     return success();
+//   }
+// };
+
+struct LowerConcatRefPass
+    : public circt::moore::impl::LowerConcatRefBase<LowerConcatRefPass> {
+  void runOnOperation() override;
+};
+
+} // namespace
+
+std::unique_ptr<mlir::Pass> circt::moore::createLowerConcatRefPass() {
+  return std::make_unique<LowerConcatRefPass>();
+}
+
+void LowerConcatRefPass::runOnOperation() {
+  MLIRContext &context = getContext();
+  ConversionTarget target(context);
+
+  target.addDynamicallyLegalOp<ConcatRefOp>([](ConcatRefOp op) {
+    auto *user = *op->getUsers().begin();
+    return !isa<BlockingAssignOp>(user);
```

Comment:
I think all concat ref operatins are illegal.

Reply:
Yeah, you're right! I also think so. However, a present, I don't how to handle the case like the instance uses the `concat_ref`:
```
module {
  moore.module @top() {
    %w1 = moore.net wire : <l1>
    %w2 = moore.net wire : <l1>
    %w3 = moore.net wire : <l1>
    %w4 = moore.net wire : <l1>
    %0 = moore.concat_ref %w1, %w2, %w3, %w4 : (!moore.ref<l1>, !moore.ref<l1>, !moore.ref<l1>, !moore.ref<l1>) -> <l4>
    moore.instance "a" @A(a: %0: !moore.ref<l4>) -> ()
    moore.output
  }
  moore.module @A(in %a : !moore.ref<l4>) {
    %a_0 = moore.net name "a" wire : <l4>
    %0 = moore.read %a : l4
    moore.assign %a_0, %0 : l4
    moore.output
  }
}
```

Reply:
Oh, I see. That's tricky. In that case you can allow concat_ref whose users are instances.

Reply:
Cool idea! I'll tweak it. Thanks! :smiley: